src/lib/blkalloc/append_blk_allocator.cpp

```diff
@@ -102,86 +80,68 @@ BlkAllocStatus AppendBlkAllocator::alloc(blk_count_t nblks, const blk_alloc_hint
+    return BlkAllocStatus::SUCCESS;
+}
 
-bool AppendBlkAllocator::is_blk_alloced_on_disk(BlkId const&, bool) const { return false; }
+BlkAllocStatus AppendBlkAllocator::alloc_on_cache(BlkId const& blkid) {
+    auto new_offset = blkid.blk_num() + blkid.blk_count();
+    auto cur_offset = m_last_append_offset.load();
+    do {
+        if (cur_offset >= new_offset) { break; } // Already allocated
+    } while (!m_last_append_offset.compare_exchange_weak(cur_offset, new_offset));
+    return BlkAllocStatus::SUCCESS;
+}
 
-//
-// cp_flush doesn't need CPGuard as it is triggered by CPMgr which already handles the reference check;
-//
-// cp_flush should not block alloc/free;
-//
 void AppendBlkAllocator::cp_flush(CP* cp) {
-    const auto idx = cp->id() % MAX_CP_COUNT;
     // check if current cp's context has dirty buffer already
-    if (m_dirty_sb[idx].is_dirty) {
-        m_sb->last_append_offset = m_dirty_sb[idx].last_append_offset;
-        m_sb->freeable_nblks = m_dirty_sb[idx].freeable_nblks;
+    if (m_is_dirty.exchange(false)) {
+        m_sb->last_append_offset = m_last_append_offset.load();
+        m_sb->freeable_nblks = m_freeable_nblks.load();
 
         // write to metablk;
         m_sb.write();
-
-        // clear this dirty buff's dirty flag;
-        clear_dirty_offset(idx);
     }
 }
 
-// updating current cp's dirty buffer context;
-void AppendBlkAllocator::set_dirty_offset(const uint8_t idx) {
-    m_dirty_sb[idx].is_dirty = true;
-
-    m_dirty_sb[idx].last_append_offset = m_last_append_offset;
-    m_dirty_sb[idx].freeable_nblks = m_freeable_nblks;
-}
-
-// clearing current cp context's dirty flag;
-void AppendBlkAllocator::clear_dirty_offset(const uint8_t idx) { m_dirty_sb[idx].is_dirty = false; }
-
 //
 // free operation does:
 // 1. book keeping "total freeable" space
 // 2. if the blk being freed happens to be last block, move last_append_offset backwards accordingly;
 //
 void AppendBlkAllocator::free(const BlkId& bid) {
-    std::unique_lock lk(m_mtx);
-    auto cur_cp = hs()->cp_mgr().cp_guard();
-    const auto n = bid.blk_count();
-    m_freeable_nblks += n;
-    if (bid.blk_num() + n == m_last_append_offset) {
-        // we are freeing the the last blk id, let's rewind.
-        m_last_append_offset -= n;
-    }
-    set_dirty_offset(cur_cp->id() % MAX_CP_COUNT);
+    auto const nblks = bid.blk_count();
+    auto exp_last_offset = bid.blk_num() + nblks;
+    auto const new_offset = m_last_append_offset - nblks;
+    if (m_last_append_offset.compare_exchange_strong(exp_last_offset, new_offset)) { return; }
+
+    // Freeing something in the middle, increment the count
+    m_freeable_nblks.fetch_add(nblks);
 }
 
+void AppendBlkAllocator::free_on_disk(BlkId const&) { m_is_dirty.store(true); }
+
 bool AppendBlkAllocator::is_blk_alloced(const BlkId& in_bid, bool) const {
     // blk_num starts from 0;
-    return in_bid.blk_num() < get_used_blks();
+    return in_bid.blk_num() < m_last_append_offset;
```

Comment:
Why not access from existing api (`get_used_nblks()`)? Isn't it better to avoid understanding the real meaning of `m_last_append_offset` that it actually means nblks, so that if in the future this field changes, only needs to update that get_used_blks(), not every where this fields is being access directly?

Reply:
I generally prefer to use member variables instead of methods to make code easier to read and provide apis only for external consumption. If you want me to revert that, I will do that.